README.md
## ![blitz.io](http://blitz.io/images/logo2.png)

Make load and performance testing a fun sport.

* Run a sprint from around the world
* Rush your API and website to scale it out
* Condition your site around the clock

## Getting started
Login to [blitz.io](http://blitz.io) and in the blitz bar type:
    --api-key

Now
    cpan Blitz
 
**In your Perl**

```Perl

    use Blitz;

    my $blitz = Blitz->new();

```

**Sprint**

```Perl

$blitz->sprint({
    url => 'www.mycoolapp.com',
    region => 'california',
    callback => \&sprint_sink($ok, $err)
});

```

**Rush**

```Perl

$blitz->rush({
    url => 'www.mycoolapp.com',
    region => 'california',
    pattern => [
        {
            start => 1,
            end => 100,
            duration => 60,
        }],
        callback => \&rush_sink($ok, $err)
});

```


Blitz/Changes
Revision history for Blitz

0.01    July 29, 2011
        First version, released on an unsuspecting world.



Blitz/lib/Blitz/API.pm
package Blitz::API;

use strict;
use warnings;

use LWP;
use Data::Dump qw(dump);
use JSON::XS;
use MIME::Base64;

=head1 NAME

Blitz::API - Perl module for API access to Blitz

=cut

sub client {
    my $this = shift;
    my $creds = shift;
    my $self = {
        credentials => $creds,
        job_id => undef,
        job_status => 'idle',
    };
    bless $self;
    return $self;
}

sub _make_url {
    my $self = shift;
    my $path = shift;
    my $url = "http://" . $self->{credentials}{host}; # fix for https later?
    $url .= ":$self->{credentials}{port}" if $self->{credentials}{port} != 80;
    $url .= $path if $path;
    print STDERR "$url\n";
    return $url;
}

sub _http_get {
    my $self = shift;
    my $path = shift;

    my $browser = LWP::UserAgent->new;
    my $url = _make_url($self, $path);
    my $response = $browser->get($url,
        'X-API-User'   => $self->{credentials}{username},
        'X-API-Key'    => $self->{credentials}{api_key},
        'X-API-Client' => 'Perl',
    );
    
    return $response;
}

=head2 _decode_result

Decode the JSON result object from Blitz
Decode base64 response in content areas of request and response

=cut

sub _decode_response {
    my $self = shift;
    my $response = shift;
    my $return = {};
    if ( $response->{_rc} != 200 ) {
        $return->{error} = 'server';
        $return->{cause} = $response->code();
    }
    else {
        $return = decode_json($response->{_content});
        for my $key ('request', 'response') {
            if ($return->{result}) {
                if ($return->{result}{$key} && $return->{result}{$key}{content}) {
                    $return->{result}{$key}{content} = decode_base64($return->{result}{$key}{content});    
                }    
            }
        }
    }
# XXX for DEBUG
#    print STDERR "----\n";
#    Data::Dump::dump($return);
#    print STDERR "----\n";

    return $return;
}

sub login {
    my $self = shift;
    my $closure = shift;
    print STDERR "\nLOGIN: \n\n";
    Data::Dump::dump($self);    
    my $response = _http_get($self, '/login/api');
    my $result = _decode_response($self, $response);

    if ($closure) {
        &$closure($self, $result);
    }
    return $result;
}

sub job_id {
    my $self = shift;
    my $job_id = shift;
    $self->{job_id} = $job_id if $job_id;
    return $self->{job_id};
}

sub job_status {

    my $self = shift;
    print STDERR "\nJOB STATUS: \n\n";
    Data::Dump::dump($self);
    my $job_id = $self->job_id;
    my $closure = $self->{callback};
    print STDERR "\nJOB STATUS: \n\n";
    Data::Dump::dump($self);

    my $request = '/api/1/jobs/' . $job_id . '/status';
    my $response = _http_get($self, $request);
    
    my $result = _decode_response($self, $response);

    if ($closure) {
        &$closure($self, $result);
    }

    print STDERR "\nJOB STATUS RETURN\n\n";
    Data::Dump::dump($result);
    return $result;
}

# client object getter/setter for api_key
# necessary?
sub api_key {
    my $self = shift;
    my $api_key = shift;
    if ($api_key) {
        $self->{credentials}{api_key} = $api_key;
    }
    return $self->{credentials}{api_key};
}

sub start_job {
    my $self = shift;
    my $data = shift;
    my $closure = $self->{callback};
    Data::Dump::dump($data);
    $data = encode_json($data);
    
    my $browser = LWP::UserAgent->new;
    my $url = _make_url($self, '/api/1/curl/execute');
    
    my $response = $browser->post($url,
        'X-API-User'     => $self->{credentials}{username},
        'X-API-Key'      => $self->{credentials}{api_key},
        'X-API-Client'   => 'Perl',
        'content-length' => length($data),
        Content          => $data,
    );
    print STDERR "\nJOB START REQUEST: \n\n";
    Data::Dump::dump($data);
    my $result = _decode_response($self, $response);
    print STDERR "\nSTART JOB: \n\n";
    Data::Dump::dump($self);
    print STDERR "\nSTART JOB RETURN\n\n";
    Data::Dump::dump($result);
    
    if ($closure) {
        &$closure($self, $result);
    }
    return $result;
}


return 1;


Blitz/lib/Blitz/Exercise.pm
package Blitz::Exercise;

use strict;
use warnings;

use Blitz;
use Blitz::API;
use Blitz::Validate;
use MIME::Base64;
use JSON::XS;
use LWP;

=head1 NAME

Blitz::Exercise - Superclass for Sprint and Rush modules

=head1 SUBROUTINES/METHODS


=head2 new

Create a blitz exercise object, which returns a closure with execute and abort methods

=cut

sub new {
    my ($name, $blitzObj, $options, $callback) = @_;
    # convenience vars
    my $self = {
        blitzObj    => $blitzObj,
        options     => $options,
        callback    => $callback,
        job_id      => 0,
    };

    bless $self;
    return $self;
}

sub blitzObj {
    my $self = shift;
    return $self->{blitzObj};
}

sub execute {
    my $self = shift;
    my $blitz = $self->blitzObj();
    my $client = Blitz::get_client($self->blitzObj());
    
    my ($valid, $result) = Blitz::Validate::validate($self->{options});
    if (!$valid) {
        &{$self->{callback}}($result, $result->{error});
    }
    else {
        # send execute request to host

        my $response = $client->start_job($self->{options}, $self->{callback});

        if ($response->{job_id}) {
            my $job_id = $client->job_id($response->{job_id});
            # wait 2 secs, then get status
            until ($client->{job_status} eq 'completed' or 
                    $client->{job_status} eq 'fail') {
                sleep 2;
                
                $response = $client->job_status();
                my $error = 0;
                if ($response->{error}) {
                    $error = $response->{error};
                }
                elsif (         
                    $response->{result} && 
                    $response->{result}{error}
                ) {
                    $error = $response->{result}{error};
                }
                if ($error) {
                    $client->{job_status} = 'fail';
                    &{$self->{callback}}($response, $error);
                }
                else {
                    $client->{job_status} = $response->{job_status};
                    &{$self->{callback}}($response, $error);
                }
            }
        }
        else {
            # no id means failure
            &{$self->{callback}}($response, 'No job id returned');
        }
    }
    return $self;
}


return 1;


Blitz/lib/Blitz/Validate.pm
package Blitz::Validate;

use strict;
use warnings;
use Regexp::Common qw/URI/;

=head1 NAME

Blitz::Validate - Perl module for assisting with the validation of calls to Blitz.io

=cut

sub validate {
    my $options = shift;
    my $test_type = shift;
    my $reasons = [];

    # url
    if (!$options->{url}) {
        push @$reasons, "No URL given";
    }
    elsif (! _is_url($options->{url}) ) {
        push @$reasons, "Invalid URL: $options->{url}";
    }

    # pattern
    if ($options->{pattern} && $test_type eq 'rush') {
        if (!$options->{pattern}{iterations} || ! _is_integer($options->{pattern}{iterations}) ) {
            push @$reasons, "Pattern iterations not given or not an integer";
        }
        if (! $options->{pattern}{intervals} || ! _is_array($options->{pattern}{intervals}) ) {
            push @$reasons, "Intervals is not an array";
        }
    }
    elsif ($options->{pattern} && $test_type eq 'sprint') {
        push @$reasons, 'Pattern is not a valid options for sprints';
    }

    # region
    if ($options->{region} && $options->{region} !~ /(california|virginia|ireland|singapore|japan)/) {

        push @$reasons, "Invalid region: $options->{region}";
    }

    # ssl
    if ($options->{ssl} && !
        ( $options->{ssl} eq 'tlsv1' || 
          $options->{ssl} eq 'sslv2' ||
          $options->{ssl} eq 'sslv3' ) 
          ){
        push @$reasons, "SSL needs to be one of the following: tlsv1, sslv2, or sslv3";
    }
    
    # data
    if ($options->{content}) {
        if (! _is_hash($options->{content})) {
            push @$reasons, "Content must be a hash";
        }
        elsif ( ! $options->{content}{data} || ! _is_array($options->{content}{data}) ) {
             push @$reasons, "Content hash must have a data key with an array value"; 
        }
    }

    # referer
    if ($options->{referer} && ! _is_url($options->{referer} )) {
        push @$reasons, "Invalid referer URL: " . $options->{referer};
    }
    # string params
    for my $key ('user-agent', 'user') {
        if ($options->{$key} && 
            ( _is_array($options->{$key}) || 
              _is_hash($options->{$key}) ) 
            ) {
                push @$reasons, "Invalid " . $key . ": " . "$options->{$key}";
        }
    }
    
    # integer params
    for my $key ('status', 'timeout') {
        if ($options->{$key} && $options->{$key} !~ /^\d+$/g) {
            push @$reasons, "Invalid " . $key . ": " . "$options->{$key}";
        }
    }
    
    # array params
    for my $key ('cookies', 'headers') {
        if ($options->{$key} && ! _is_array($options->{$key})) {
            push @$reasons, "values to $key not given as an array: " . $options->{$key};
        }
    }
    
    # XXX: variables
    # variables must be a hash
    # key/val pairs of hash must be name, followed by a hash of params
    # each params hash must have a type key, with the value being one of "list|alpha|number|udid"
    # if list, must have an entries key with a value of an array
    # if alpha, must have keys min and max, with integer values
    # if number, must have keys min and max, with integer values
    if ($options->{variables}) {
        if (! _is_hash($options->{variables})) {
            push @$reasons, "values to variables not given as a hash: " . $options->{variables};
        }
        else {
            for my $key (keys %{$options->{variables}}) {
                if (! _is_hash($options->{variables}{$key})) {
                    push @$reasons, 
                        "values to variables key $key not given as a hash: " . 
                        $options->{variables}{$key};
                }
                else {
                    my $type = $options->{variables}{$key}{type};
                    if (! $type) {
                        push @$reasons, "all variables must have a type definition";
                    }
                    elsif (! $type =~ /^(list|alpha|number|udid)$/) {
                        push @$reasons, "all variables must have type of list, alpha, number, or udid";
                    }
                    else {
                        if ($type eq 'list') {
                            if (! $options->{variables}{$key}{entries} ||
                                ! _is_array($options->{variables}{$key}{entries}) ) {
                                    push @$reasons, "all variables of type list need an array of values";
                            }
                        }
                        elsif ($type eq 'number' || $type eq 'alpha') {
                            for my $int ('min', 'max') {
                                if (! $options->{variables}{$key}{$int} ||
                                    $options->{variables}{$key}{$int} !~ /^\-*\d+$/g) {
                                        push @$reasons, "$int value must be an integer for $key";
                                }
                            }
                        }
                    }
                }
            }
        }
    }
    
    if ($reasons->[0]) {
        my $reason = join(', ', @$reasons);
        return (0, { error => 'Invalid parameters', reason => $reason });
    }
    else {
        return (1, {});
    }
}

sub _is_integer {
    my $val = shift;
    if ($val && $val =~ /^\d+$/g) {
        return 1;
    }
    else {
        return 0;
    }
}

sub _is_foo {
    my ($val, $type) = @_;
    if ( ref($val) =~ $type) {
        return 1;
    }
    else {
        return 0;
    }
}
sub _is_array {
    return _is_foo($_[0], 'ARRAY');
}

sub _is_hash {
    return _is_foo($_[0], 'HASH');
}

sub _is_url {
    my $url = shift;
    return $url =~ /$RE{URI}{HTTP}{ -scheme => 'https?' }/g;
}

return 1;


Blitz/lib/Blitz.pm
package Blitz;

use strict;
use warnings;

use Blitz::API;
use Blitz::Exercise;
use Blitz::Sprint;
use Blitz::Rush;

=head1 NAME

Blitz - Perl module for API access to Blitz

=head1 VERSION

Version 0.01

=cut

our $VERSION = '0.01';

=head1 SYNOPSIS

Blitz provides an interface to the blitz API. Blitz is a
performance and load testing application for testing cloud service apps. 
More information on blitz can be found at http://blitz.io


    use Blitz;

    my $blitz = Blitz->new;

=cut

=head1 SUBROUTINES/METHODS

=head2 new

# Create a new Blitz object

    my $blitz = Blitz->new;

=cut

sub new {
    my $name = shift;
    my $this = shift || {};
    my $self = {
        _authenticated => 0,
        credentials => {
            username => $this->{username},
            api_key   => $this->{api_key},
            host     => $this->{host} || 'blitz.io',
            port     => $this->{port} || 80,
        }
    };
    bless $self;
    return $self;
}

sub _get_set_credentials {
    my $self = shift;
    my $field = shift;
    my $value = shift;
    if ($value) {
        $self->{credentials}{$field} = $value;
    }
    return $self->{credentials}{$field};
}

=head2 username

# Get the currently configured username

    my $user = $blitz->username;

# Set the username

    my $user = $blitz->username('joe@joe.org');

=cut

sub username {
    my $self = shift;
    return _get_set_credentials($self, 'username', @_);
}

=head2 api_key

# Get the currently configured api_key

    my $api_key = $blitz->api_key;

# Set the api_key

    my $api_key = $blitz->api_key('706d5cfbd3338ba110bfg6f46d91f8f3');

=cut

sub api_key {
    my $self = shift;
    return _get_set_credentials($self, 'api_key', @_);
}


=head2 host

# Get the currently configured host

    my $host = $blitz->host;

# Set the host

    my $host = $blitz->host('foo.com');

=cut

sub host {
    my $self = shift;
    return _get_set_credentials($self, 'host', @_);
}

=head2 port

# Get the currently configured port

    my $port = $blitz->port;

# Set the host

    my $port = $blitz->port(8080);

=cut

sub port {
    my $self = shift;
    return _get_set_credentials($self, 'port', @_);
}

=head2 authenticated

Have we been authenticated?

    my $auth = $blitz->authenticated;

=cut

sub authenticated {
    my $self = shift;
    return $self->{_authenticated};
}

=head2 get_client

fetches the existing client object, or 
creates a new Blitz::API->client object

    my $client = $blitz->get_client;
    
=cut

sub get_client {
    my $self = shift;
    if (! $self->{client}) {
        my $client = Blitz::API->client($self->{credentials});
        $self->{client} = $client;
    }
    return $self->{client};
}

sub _run {
    my ($self, $obj, $options, $callback) = @_;
    if ($self->{_authenticated}) {
        my $exercise = $obj->new(
                $self,
                $options,
                $callback
            );
        $exercise->execute();
    }
    else {
        my $client = $self->get_client;
        $client->login(
            sub { 
                my $self = shift;
                my $result = shift;
                if ($result->{ok}) {
                    $self->{_authenticated} = 1;
                    $self->{credentials}{api_key} = $result->{api_key};
                    my $exercise = $obj->new(
                        $self,
                        $options, 
                        $callback
                        );
                    $exercise->execute();
                }
                else {
                    &$callback($result, $result->{error});
                }
            }
        );
    }
}


=head2 sprint

# Sprint

    $blitz->sprint(
        {
            url => 'www.mycoolapp.com',
            region => 'california',
            },
            callback
        }
    );


=cut

sub sprint {
    my $self = shift;
    my $options = shift;
    my $callback = shift;
    _run($self, 'Blitz::Sprint', $options, $callback);
}

=head2 rush

# Rush

    $blitz->rush(
        {
            url => 'www.mycoolapp.com',
            region => 'california',
            pattern => [
                {
                    start => 1,
                    end => 100,
                    duration => 60,
                }
            ]
        },
        callback
        }
    );


=cut


sub rush {
    my $self = shift;
    my $options = shift;
    my $callback = shift;
    _run($self, 'Blitz::Rush', $options, $callback);
}

=head1 AUTHOR

Ben Klaas, C<< <ben at benklaas.com> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-blitz at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Blitz>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.




=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Blitz


You can also look for information at:

=over 4

=item * Github: Open source code repository

L<http://github.com/bklaas/blitz-perl>

=item * RT: CPAN's request tracker (report bugs here)

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Blitz>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Blitz>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Blitz>

=item * Search CPAN

L<http://search.cpan.org/dist/Blitz/>

=back


=head1 ACKNOWLEDGEMENTS

Thanks to Guilherme Hermeto and Kowsik Guruswamy for assistance
in understanding the blitz API and requirements of the Perl client.

=head1 LICENSE AND COPYRIGHT

This software is under the MIT license

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.

=cut

1; # End of Blitz
